### PR TITLE
fix(sdk): avoid unlinking QE library when running "prisma generate"

### LINKED ISSUE #9184: EPERM: operation not permitted, unlink '...node_modules\prisma\query_engine-windows.dll.node'
### Bug description

Following the [Getting Started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) guide, when it gets to the point of running the "migrate dev" command, the script tries to install some packages through npm (don't know which ones) and this causes the error in title.

### How to reproduce

1. Follow the [Getting started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) prisma guide up to and included the "npx prisma migrate dev --name init" command.
2. Get the following error:

```
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "testprisma", schema "public" at "localhost:5432"

PostgreSQL database testprisma created at localhost:5432

The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20210909080511_init/
    └─ migration.sql

Your database is now in sync with your schema.

Running generate... (Use --skip-generate to skip the generators)
npm WARN hello-prisma@1.0.0 No description
npm WARN hello-prisma@1.0.0 No repository field.

npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node'
npm ERR!   },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node',
npm ERR!   parent: 'hello-prisma'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\mbelletti\AppData\Roaming\npm-cache\_logs\2021-09-09T08_05_23_557Z-debug.log
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.1
The batch file cannot be found.
```

The log file is the following:
[2021-09-09T08_05_23_557Z-debug.log](https://github.com/prisma/prisma/files/7134677/2021-09-09T08_05_23_557Z-debug.log)

After this, the node_modules\.bin directory no longer has the prisma executable, and the node_modules/prisma folder only contains the file that npm wasn't able to remove (query_engine-windows.dll.node).

### Expected behavior

The script finishes correctly.

### Prisma information

All the prisma configurations are the ones taken from the guide.

### Environment & setup

- OS: Windows 10
- Database: PostgreSQL 13
- Node.js version: 14.17.1 (using nvm)

### Prisma Version

```
Environment variables loaded from .env
prisma                  : 3.0.1
@prisma/client          : Not found
Current platform        : windows
Query Engine (Node-API) : libquery-engine 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 2452cc6313d52b8b9a96999ac0e974d0aedf88db
Studio                  : 0.423.0
```

(I had to relaunch npm install to get prisma back)

### PR DESCRIPTION
**Pull request by aqrln**
* Don't install `prisma` when it is already present.
* Don't install both `prisma` and `@prisma/client` if Yarn on Windows is detected, and ask the user to install them manually.
* Add TODO comments for potential bugs out of scope of this PR.

Fixes https://github.com/prisma/prisma/issues/9184

### CODE REVIEW

**Comment by aqrln** [comment]
Marking as draft for now because some testing is necessary

File: packages/sdk/src/predefinedGeneratorResolvers.ts [hunk 1]
```diff
@@ -75,6 +75,8 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
     await checkTypeScriptVersion()
 
     if (!prismaClientDir && !process.env.PRISMA_GENERATE_SKIP_AUTOINSTALL) {
+      // TODO: `prisma generate` may be called deeper than one subdirectory from
+      // the package root.
```
Comment (janpio) [inline packages/sdk/src/predefinedGeneratorResolvers.ts:79]: What is the potential impact of this?

File: packages/sdk/src/predefinedGeneratorResolvers.ts [hunk 2]
```diff
@@ -97,7 +99,30 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
...
+            `${chalk.bold.greenBright(`${getAddPackageCommandName(baseDir, 'dev')} prisma`)} and `,
+          )}${chalk.bold.greenBright(`${getAddPackageCommandName(baseDir)} @prisma/client`)}, and rerun ${chalk.bold(
+            getCommandWithExecutor('prisma generate'),
+          )} 🙏.`,
+        )
+      }
+
+      if (!prismaCliDir) {
+        await installPackage(baseDir, `prisma@${version ?? 'latest'}`, 'dev')
+      }
```
Comment (janpio) [inline packages/sdk/src/predefinedGeneratorResolvers.ts:124]: Does this make sense to have this inside the check in line 77? I think we could move it outside so it is also run if `@prisma/client` already exists.

Comment (janpio) [reply packages/sdk/src/predefinedGeneratorResolvers.ts:124]: Oh ok, I see lines 104 to 119 would also need to be refactored for that to work.

**Review by janpio — COMMENTED** [review]
Looks fundamentally good to me.

File: packages/sdk/src/predefinedGeneratorResolvers.ts (hunk 1 above)
Comment (aqrln) [reply packages/sdk/src/predefinedGeneratorResolvers.ts:79]: The potential impact is creating a phony `package.json` file somewhere in the project. Shouldn't actually break anything if that happens, but could be annoying.

**Comment by aqrln** [comment]
Manual testing ✅

How should we test this automatically? e2e tests?

File: packages/sdk/src/predefinedGeneratorResolvers.ts (hunk 1 above)
Comment (janpio) [reply packages/sdk/src/predefinedGeneratorResolvers.ts:79]: Ok I see. Yeah, that is not optimal.

**Comment by janpio** [comment]
e2e-tests repo would be the fallback, yes, if we can not do that in the tests here. As it does not need to talk to the database, but really just is about running a few commands in a specific order it does not _really_ fit there either but we should be ok with that.

**Review by Jolg42 — APPROVED** [review]
That will make some people quite happy :)

**Comment by janpio** [comment]
Oh, this is an intergration branch so I can actually test it on Windows 👍  Let me do that...

**Review by janpio — APPROVED** [review]

**Review by millsp — APPROVED** [review]

**Pull request merged**